Implemented redaction functions to obscure sensitive information in strings

abhishek9sharma opened the pull request:
- added below functions in common.py
   - [redact]
   - [ismatchingkey](https://github.com/guardrails-ai/guardrails/compare/main...abhishek9sharma:guardrails:main#diff-17ac5d80046e0124d7818ae1e1c56afe9ce6f7a9211fb039d63fa801953bca3dR144)
   - [can_convert_to_dict](https://github.com/guardrails-ai/guardrails/compare/main...abhishek9sharma:guardrails:main#diff-17ac5d80046e0124d7818ae1e1c56afe9ce6f7a9211fb039d63fa801953bca3dR165)
   - [recursive_key_operation](https://github.com/guardrails-ai/guardrails/compare/main...abhishek9sharma:guardrails:main#diff-17ac5d80046e0124d7818ae1e1c56afe9ce6f7a9211fb039d63fa801953bca3dR185)
   
   - Fix for the bug. https://github.com/guardrails-ai/guardrails/issues/1213

abhishek9sharma committed:
- Implemented redaction functions to obscure sensitive information in strings,
- including `redact`, `ismatchingkey`, `can_convert_to_dict`, and
- `recursive_key_operation`.
- Added unit tests for the new functionalities to ensure correctness and reliability.

abhishek9sharma committed:
fix docstring

abhishek9sharma committed:
- Refactored unit tests for recursive key operations

abhishek9sharma committed:
Merge pull request #1 from abhishek9sharma/bugfix/1213

- Implemented redaction functions to obscure sensitive information in…

abhishek9sharma committed:
- Added recursive key operation and redaction for invocation params

abhishek9sharma committed:
Merge pull request #2 from abhishek9sharma/bugfix/1213

- Added recursive key operation and redaction for invocation params

CalebCourier commented on guardrails/telemetry/open_inference.py:
```diff
@@ -93,6 +99,9 @@ def trace_llm_call(
+        )
         current_span.set_attribute(
             "llm.invocation_parameters", ser_invocation_parameters
         )
```
```suggestion
    redacted_ser_invocation_parameters = recursive_key_operation(
        ser_invocation_parameters, redact
    )
    reser_invocation_parameters = (
        json.dumps(redacted_ser_invocation_parameters)
        if isinstance(redacted_ser_invocation_parameters, dict)
        or isinstance(redacted_ser_invocation_parameters, list)
        else redacted_ser_invocation_parameters
    )
    if reser_invocation_parameters:
        current_span.set_attribute(
            "llm.invocation_parameters", reser_invocation_parameters
        )
```

CalebCourier replied:
@abhishek9sharma overall looks good! I'm making this suggestion to appease the type-checking.

abhishek9sharma replied:
@CalebCourier  thanks for the comments. Do you want  me to add checks in this file. I fixes some missing hints for the functions being called here https://github.com/guardrails-ai/guardrails/pull/1217/commits/65dcf5af1ecf47692fa2d32dc5022735bad0e431

CalebCourier replied:
No, I think your changes along with these should do the trick.  I'm going to commit the suggestion and re-run the checks.

CalebCourier commented:
Tests are passing locally.  Codecov generally fails on PRs from forks because of auth, so disregard those failed checks.

abhishek9sharma committed:
- Add type hints to function signatures in `common.py` for improved
- code clarity and type safety.
- Updated `can_convert_to_dict` and `recursive_key_operation` functions
- to specify input and output types.

abhishek9sharma committed:
Merge pull request #3 from abhishek9sharma/bugfix/1213

- Add type hints to function signatures in `common.py` for improved

CalebCourier committed:
Reserialize redacted parameters to strings

CalebCourier commented on guardrails/telemetry/open_inference.py:
```diff
@@ -1,6 +1,12 @@
 from typing import Any, Dict, List, Optional
```
```suggestion
import json
from typing import Any, Dict, List, Optional
```

CalebCourier committed:
Update guardrails/telemetry/open_inference.py

CalebCourier commented on guardrails/telemetry/common.py:
```diff
@@ -124,3 +124,94 @@ def add_user_attributes(span: Span):
+
+
+def recursive_key_operation(
+    data: Dict[str, Any] | List[Any] | str,
```
```suggestion
    data: Optional[Union[Dict[str, Any], List[Any], str]],
```

CalebCourier replied:
We still support down to python 3.9 so we have to use `Union` instead of `|` for now.

CalebCourier commented on guardrails/telemetry/common.py:
```diff
@@ -124,3 +124,94 @@ def add_user_attributes(span: Span):
+    data: Dict[str, Any] | List[Any] | str,
+    operation: Callable[[str], str],
+    keys_to_match: List[str] = ["key", "token"],
+) -> Dict[str, Any] | List[Any] | str:
```
```suggestion
) -> Optional[Union[Dict[str, Any], List[Any], str]]:
```

CalebCourier replied:
We still support down to python 3.9 so we have to use `Union` instead of `|` for now.

CalebCourier commented on guardrails/telemetry/common.py:
```diff
@@ -124,3 +124,94 @@ def add_user_attributes(span: Span):
+        data = str(recursive_key_operation(data_dict, operation, keys_to_match))
+    elif isinstance(data, dict):
+        for key, value in data.items():
+            if ismatchingkey(key, keys_to_match) and isinstance(value, str):
```
```suggestion
            if ismatchingkey(key, tuple(keys_to_match)) and isinstance(value, str):
```

CalebCourier committed:
cast keys_to_match to tuple

CalebCourier committed:
Use older Union syntax for older python version support

CalebCourier committed:
Use older Union syntax for older python version support

CalebCourier commented on guardrails/telemetry/common.py:
On last change, we need to add `List` to the imports from `typing` at the top.  Github won't let me make a suggestion that far up since there weren't changes there.

abhishek9sharma committed:
- Fix imports
- Updated docstring for clarity and improved input handling.

abhishek9sharma committed:
Merge pull request #4 from abhishek9sharma/bugfix/1213

- Fix imports

CalebCourier commented:
@abhishek9sharma Thanks again for your contribution! This will be part of our release next week.  I also filed an issue concerning the `Call` logs you mentioned here: https://github.com/guardrails-ai/guardrails/issues/1221

abhishek9sharma commented:
@CalebCourier  I can try the new issue sometime later this week. Also added a PR to log span_kind. Let me know if it does not make sense. https://github.com/guardrails-ai/guardrails/pull/1223
